### PR TITLE
fix: Ensure each table is processed in its own `LogicalPlan` graph

### LINKED ISSUE #7794: InfluxQL-IOx: Inconsistent output for aggregate query with multiple measurements or subqueries in `FROM`
Aggregate queries over multiple data sources in the `FROM` clause return incorrect results.

# Background

With the current implementation of the InfluxQL planner, the order of operations for projecting a `SELECT` with aggregate expressions over multiple tables in the `FROM` clause is:

```
┌───────────┐   ┌───────────┐                                                
│  FROM 1   │──▶│ Aggregate │─┐                                              
└───┬───┬───┘   └───────────┘ │ ┌───────────┐   ┌───────────┐   ┌───────────┐
    │...│                     ├▶│   Union   │──▶│   Sort    │──▶│   Limit   │
┌───┴───┴───┐   ┌───────────┐ │ └───────────┘   └───────────┘   └───────────┘
│  FROM n   │──▶│ Aggregate │─┘                                              
└───────────┘   └───────────┘                                                
```

The following InfluxQL query would produce the previously described plan:

```SQL
SELECT last(usage_idle), last(usage_system) FROM (SELECT usage_idle FROM cpu WHERE cpu = 'cpu-total'), (SELECT usage_system FROM cpu WHERE cpu = 'cpu-total')
```

The expected output is, noting that the values for the same measurement and timestamp produce a single row:

```text
name: cpu
+---------------------+------+--------+
| time                | last | last_1 |
+---------------------+------+--------+
| 1970-01-01T00:00:00 | 2.99 | 2.1    |
+---------------------+------+--------+
```

however, as the aggregates are performed first, followed by the union, the output produced by IOx is:

```text
name: cpu
+---------------------+------+--------+
| time                | last | last_1 |
+---------------------+------+--------+
| 1970-01-01T00:00:00 | 2.99 |        |
| 1970-01-01T00:00:00 |      | 2.1    |
+---------------------+------+--------+
```

> **Note**
>
> Different measurements (tables) are never combined, even when using subqueries.

It is possible to work around the issue by using a single data source in the `FROM` clause:

```SQL
SELECT last(usage_idle), last(usage_system) FROM (SELECT usage_idle, usage_system FROM cpu WHERE cpu = 'cpu-total');
```

# Solution

The solution is to combine the inputs with the UNION operator followed by the aggregate.

```text
┌───────────┐                                                                
│  From 1   │─┐                                                              
└───┬───┬───┘ │ ┌───────────┐   ┌───────────┐   ┌───────────┐   ┌───────────┐
    │...│     ├▶│   Union   │──▶│ Aggregate │──▶│   Sort    │──▶│   Limit   │
┌───┴───┴───┐ │ └───────────┘   └───────────┘   └───────────┘   └───────────┘
│  FROM n   │─┘                                                              
└───────────┘                                                                
```

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #7794

The InfluxQL planner now generates a separate plan graph for each unique table, which more closely models the InfluxQL OG query engine. This was necessary so that aggregate expressions could be applied to the union of all projections for the same measurement, ensuring that rows for the same timestamp were coalesced together:

https://github.com/influxdata/influxdb_iox/blob/50289ebab75e17ef5ab4a56ced46ef502f3bfe9e/influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql.expected#L2926-L2933

whereas previously, the prior query produced the following, incorrect output:

```
+---------------------+------+--------+
| time                | last | last_1 |
+---------------------+------+--------+
| 2022-10-31T02:00:00 | 2.98 |        |
| 2022-10-31T02:00:00 |      | 2.2    |
| 2022-10-31T02:00:10 | 2.99 |        |
| 2022-10-31T02:00:10 |      | 2.1    |
+---------------------+------+--------+
```

Altering the plan affected the logical plans for any `SELECT` statement that specified a subquery, as can be observed by the changes to the `insta` test snapshots. Whereas previously, the planner was injecting the `iox::measurement_name` column from the first projection:

https://github.com/influxdata/influxdb_iox/blob/c082ebac7dc056f5da860cc58863ab602d7016e8/iox_query_influxql/src/plan/planner.rs#L2639-L2640

it is now only necessary to project it at the end, to ensure rows for each measurement are separate in the final output:

https://github.com/influxdata/influxdb_iox/blob/50289ebab75e17ef5ab4a56ced46ef502f3bfe9e/iox_query_influxql/src/plan/planner.rs#L2809-L2810

This benefit reduces the number of columns and thus the complexity of the sort until the outermost query.

### CODE REVIEW

File: iox_query_influxql/src/plan/planner.rs [hunk 1]
```diff
@@ -730,8 +766,7 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
     /// - `offset`: The number of input rows to skip.
     /// - `limit`: The maximum number of rows to return in the output plan per group.
     /// - `time_sort_expr`: An `Expr::Sort` referring to the `time` column of the input.
-    /// - `has_multiple_measurements`: `true` if the `input` produces multiple measurements,
-    ///   and therefore the limit should be applied per measurement and any additional group tags.
+    /// - `sort_by_measurement`: `true` if the `input` must be sorted by the measurement column.
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:769]: Renamed this for clarity

File: iox_query_influxql/src/plan/planner/select.rs [hunk 2]
```diff
@@ -80,16 +76,7 @@ pub(super) fn plan_with_sort(
...
     ) -> impl Iterator<Item = Expr> + 'a {
-        fields
-            .iter()
-            .filter(|f| {
-                if let Ok(df) = schema.field_with_unqualified_name(f) {
-                    *df.data_type() != DataType::Null
-                } else {
-                    false
-                }
-            })
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner/select.rs:91]: Replaced with `fields_to_exprs_no_nulls`, which was a duplicate of this code.

**Review by stuartcarnie — COMMENTED** [review]
Added some comments

**Review by mhilton — APPROVED** [review]
LGTM. I don't really have enough context on datafusion to know if there is a potential problem with the plans being created, if you are concerned about that then maybe wait for a second review.

File: influxdb_influxql_parser/src/expression/arithmetic.rs [hunk 3]
```diff
@@ -19,7 +19,7 @@ use std::fmt::{Display, Formatter, Write};
 use std::ops::Neg;
 
 /// Reference to a tag or field key.
-#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
```
[the remark this answers is not in the document]
Comment (wolffcm) [reply influxdb_influxql_parser/src/expression/arithmetic.rs:22]: I think if Clippy doesn't complain about it, there's no convention ;) But I normally alphabetize them as well.

File: influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql [hunk 4]
```diff
@@ -568,70 +568,47 @@ SELECT P / 99.99 FROM (SELECT max(usage_system) AS P, mean(usage_idle) AS S FROM
...
--- Similar query produces expected output of prior query
+-- Multiple subqueries from same measurement, GROUP BY TIME
 SELECT last(sum_idle), last(sum_system) FROM (SELECT sum(usage_idle) AS sum_idle, sum(usage_system) AS sum_system FROM cpu WHERE cpu = 'cpu-total' GROUP BY time(10s)) GROUP BY time(10s) FILL(none);
 
--- TODO(sgc): https://github.com/influxdata/influxdb_iox/issues/7794
+-- Select the last value from two subqueries using two fields
+-- aliased to the same name, `sum_idle`. This results in multiple
+-- multiple rows for each timestamp.
+--
+-- Result  : Inconsistent
```
Comment (alamb) [inline influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql:589]: Maybe it is worth adding a link to the relevant ticket that tracks the inconsistent results 🤔

File: iox_query_influxql/src/plan/planner.rs [hunk 5]
```diff
@@ -2207,10 +2269,53 @@ fn add_time_restriction(plan: LogicalPlan, cutoff: MetadataCutoff) -> Result<Log
     }
 }
 
+/// Find distinct occurrences of `Expr::VarRef` expressions for
+/// the `select`.
+fn find_var_refs(select: &Select) -> Vec<&VarRef> {
+    let mut var_refs = Vec::new();
```
Comment (alamb) [inline iox_query_influxql/src/plan/planner.rs:2275]: I doubt it matters performance wise, but a `BTreeSet` would might make this code less verbose as it would avoid the checks for contains and sort.

File: iox_query_influxql/src/plan/rewriter.rs [hunk 6]
```diff
@@ -33,32 +35,28 @@ pub(super) fn rewrite_statement(
...
 }
 
-/// Determines if s projects more than a single unique table
-fn has_multiple_measurements(s: &Select) -> bool {
+/// Find the unique list of tables used by `s`, recursively following all `FROM` clauses and
+/// return the results in lexicographically in ascending order.
+pub(super) fn find_table_names(s: &Select) -> Vec<&str> {
     let mut data_sources = vec![s.from.as_slice()];
-    let mut table_name: Option<&str> = None;
+    let mut tables = Vec::new();
```
Comment (alamb) [inline iox_query_influxql/src/plan/rewriter.rs:45]: Again, BTreeSet might be helpful here

**Review by alamb — APPROVED** [review]
I looked at the code and it looked well structured and tested to me 🦾 

I also looked at the tests and plan changes and they all looked good to me. I am not an expert in the InfluxQL frontend but this seems like an improvement to me. Nice work @stuartcarnie

File: iox_query_influxql/src/plan/planner.rs (hunk 5 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:2275]: Nerd-sniped me to validate my previous assumptions derived from Go and other languages, which was to favour slices over maps / sets when the number of elements was normally quite small. I ran some criterion benchmarks to revalidate.

## TL;DR

They both performed within nanoseconds of each other for 3 and 13 elements. I'll change to `BTreeSet` to favour simplifying the code.

## Background

Read up on `BTreeSet`, which refers you to reading `BTreeMap`, given set is just a map with an empty type for the values. Some relevant excerpts from the docs were:

> A B-Tree instead makes each node contain B-1 to 2B-1 elements in a contiguous array. By doing this, we reduce the number of allocations by a factor of B, and improve cache efficiency in searches.

I looked up what $B$ is, and it is 6.

> Currently, our implementation simply performs naive linear search. This provides excellent performance on small nodes of elements which are cheap to compare.

That means a small `BTreeSet` has similar characteristics to a small vector.

### Benchmarks

**3 unique elements**

This first chart is for `Vec` with a `SELECT` statement that should produce 3 unique `VarRefs`:

CleanShot 2023-05-23 at 09 03 36@2x

compared to the `BTreeSet` with the same `SELECT`:

CleanShot 2023-05-23 at 09 06 47@2x

The median and mean iteration times between the algorithms is negligible.

**13 unique elements**

Once again `Vec` first:

CleanShot 2023-05-23 at 09 08 46@2x

vs `BTreeSet`:

CleanShot 2023-05-23 at 09 10 19@2x

The median and mean iteration times between the algorithms narrows.

File: iox_query_influxql/src/plan/rewriter.rs (hunk 6 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/rewriter.rs:45]: Per https://github.com/influxdata/influxdb_iox/pull/7808#discussion_r1201282970, I will update this also.

Thanks @alamb

File: influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql (hunk 4 above)
Comment (stuartcarnie) [reply influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql:589]: I was conflicted about how to track these yesterday, as I didn't want to create issues for each inconsistency that was considered low priority as this one in particular is somewhat nonsensical. I like the idea of a single ticket that summarise the inconsistencies. If we want, we can turn specific ones into issues and link them to the main issue.

File: influxdb_influxql_parser/src/expression/arithmetic.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply influxdb_influxql_parser/src/expression/arithmetic.rs:22]: Early on, I tried sorting them too, but gave up on that as `clippy` didn't care. I like the pseudo groupings that CLion does, such as the ordering and equality traits together.
